File: LeetCode/025. Reverse Nodes in k-Group/solution.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        def check(cur, k):
            for _ in range(k):
                if not cur:
                    return False
                cur = cur.next
            return True
        def reverse(cur, pre, k):
            for _ in range(k):
                nxt = cur.next
                cur.next = pre
                pre = cur
                cur = nxt
            return cur, pre
        dummy = ListNode()
        dummy.next = head
        pre, cur = dummy, head
        while cur:
            if not check(cur, k):
                break
            tmp = pre
            cur, pre = reverse(cur, pre, k)
            tmp.next.next = cur
            tmp.next, pre = pre, tmp.next
        return dummy.next

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def rev(self, h):
        tmp = h.next
        if not tmp:
            return h
        new_h = self.rev(tmp)
        tmp.next = h
        h.next = None
        return new_h
    def reverseKGroup(self, head: ListNode, k: int) -> ListNode:
        cur = head
        pre = None
        for i in range(k):
            if cur == None:
                return head
            pre = cur
            cur = cur.next
        pre.next = None
        new_head = self.rev(head)
        head.next = self.reverseKGroup(cur, k)
        return new_head

File: LeetCode/025. Reverse Nodes in k-Group/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        ListNode *dummy = new ListNode();
        ListNode *pre = dummy, *cur = head, *nxt, *f, *t;
        dummy->next = head;
        while(1){
            t = cur;
            for (int i = 0; i < k-1 && t; i++) t = t->next;
            if (!t)
                return dummy->next;
            t = t->next;
            f = pre;
            // cout << cur->val << " " << f->val << endl;
            for (int i = 0; i < k; i++){
                nxt = cur->next;
                cur->next = pre;
                pre = cur;
                cur = nxt;
            }
            f->next->next = t;
            swap(f->next,pre);
        }
        return nullptr;
    }

};




/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        if (!head)
            return head;
        // cout << head << " " << head->val  << endl;
        auto [h,t] = reverseK(head,k);
        if (!h)
            return head;
        // cout << head << " " << head->val << " " << h->val << endl;
        head->next = reverseKGroup(t,k);
        return h;
    }

    pair<ListNode*,ListNode*> reverseK(ListNode* head, int k){
        if (!head)
            return {nullptr, nullptr};
        if (k == 1)
            return {head, head->next};
        
        ListNode *nxt = head->next;
        auto [h,t] = reverseK(head->next,k-1);
        if (!h)
            return {h,t};
        nxt->next = head;
        return {h,t};
    }
};